stm32 LL driver: F1/F4 (v1 controller), F3/F7/LX (v2 controller)

ldts opened the pull request:
This patchset replaces the current stm32lx i2c driver with a generic implementation covering all the STM32 controllers.

@ydamigos implemented support for the v1 controller.
@erwango defined the driver requirements and initial tests.

F4X (v1)  has been tested on Nucleo F401re using the Grove LCD RGB sample code in Zephyr in polling/IRQ mode.
The F3 and LX (v2) family were also validated.

sidcha committed:
arm: stm32f1: Add support for STM32F103x8 SoC

Add configuration and memory definitions to support STM32F103x8
Medium-density performance line SoC with 64 KB Flash.

Merge multiple files into single Kconfig.defconfig.stm32f103xx

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

sidcha committed:
boards: arm: Add support for STM32 Minimum Development Board

Add configuration, documentation, pinmux, fixup and dts support for
STM32F103x8 based Minimum System Development board.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

jamike committed:
boards: nucleo_l476rg: enable MPU

Signed-off-by: Michel Jaouen <michel.jaouen@st.com>

jamike committed:
arm: soc: stm32: make mpu f4 config useable for other family

As other stm32 series support MPU, move common file in a file tree
useable by socs from other series

JIRA: ZEP-2220

Signed-off-by: Michel Jaouen <michel.jaouen@st.com>

jamike committed:
arm: soc: stm32: l4: add MPU capability for series

Since all l4 socs support MPU, add MPU capability in KConfig.series

JIRA: ZEP-2220

Signed-off-by: Michel Jaouen <michel.jaouen@st.com>

jamike committed:
boards: disco_l475_iot1: enable MPU

Signed-off-by: Michel Jaouen <michel.jaouen@st.com>

ydamigos committed:
pinmux: stm32f4: Add I2C1, I2C2 and I2C3 pins

Add I2C1 pins on PB6, PB7, PB8, PB9
Add I2C2 pins on PB3, PB10, PB11, PF0, PF1, PH4, PH5
Add I2C3 pins on PA8, PB4, PC9, PH7, PH8

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>
Reviewed-by: Jorge Ramirez-Oritz <jorge.ramirez-ortiz@linaro.org>

ydamigos committed:
pinmux: stm32: 96carbon: Add support for I2C

Add I2C support for Carbon from 96Boards.

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>
Reviewed-by: Jorge Ramirez-Ortiz <jorge.ramirez-ortiz@linaro.org>

galak committed:
boards: 96b_nitrogen: Add support for flash/debug with pyOCD

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: bossa-flash.sh: fix variable usage

We define a variable to pickup a default for the bossa binary, however
we weren't using it.  Lets do so now.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

ldts committed:
drivers: i2c: stm32 LL F1/F4 (v1) STM32 F3/L4X(v2)

Supports STM32 F1/F4 (v1 controller) and STM32 F3/L4X (v2
controller)

v1 could also support L1X.
v2 could also support F7X.

Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez-ortiz@linaro.org>
Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>
Reviewed-by: Yannis Damigos <giannis.damigos@gmail.com>
Tested-by: Yannis Damigos <giannis.damigos@gmail.com>
Tested-by: Jorge Ramirez-Ortiz <jorge.ramirez-ortiz@linaro.org>

ldts committed:
boards: disco_l475: i2c LL driver

Replace native driver with the LL driver

Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez-ortiz@linaro.org>

ldts committed:
pinmux: stm32: nucleo_f401re: Add support for I2C1

Signed-off-by Jorge Ramirez-Ortiz <jorge.ramirez-ortiz@linaro.org>

ldts committed:
ext: hal: stm32f1x and stm32f4x: disable i2c HAL

Enablign the HAL and the LL flags duplicate definitions when
compiling in driver. Since the i2c HAL is not needed, disable it.

This is a HAL issue that needs to be addressed (not zephyr)

Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez-ortiz@linaro.org>

mnkp committed:
boards: sam_e70_xplained: allow flashing via JTAG header

Allow to use an external debug adapter such as J-Link or ULINK
connected to a 20-pin JTAG header to flash the image. SWD is
the actual protocol used by the debug interface.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

galak committed:
scripts: pyocd.sh: Add support for passing board_id to pyocd commands

If we have more than one board of a given type we need to be able to
specify the board_id to select which specific board that the pyocd
command should target.  Introduce an environment variable PYOCD_BOARD_ID
to we can set that will get passed to the pyocd command that needs it.

Here's an example:

$ make -C samples/hello_world/ BOARD=frdm_k64f flash PYOCD_BOARD_ID=1234

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
board: frdm_k64f: allow overriding default debug/flash scripts

As there are multiple ways to flash or debug (pyOCD or openOCD) allow
the user to override the default.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented on arch/arm/soc/st_stm32/stm32f4/soc_pinmux.c:
```diff
@@ -114,6 +114,8 @@ static const stm32_pin_func_t pin_pa3_funcs[] = {
 
 static const stm32_pin_func_t pin_pa8_funcs[] = {
 	PINMUX_UART(PA8, UART7, RX)
+	[STM32F4_PINMUX_FUNC_PA8_I2C3_SCL - 1] =
```
Can we not define a macro similar to PINMUX_UART()?

ldts replied:
OK

galak commented on drivers/i2c/Kconfig:
```diff
@@ -90,19 +90,36 @@ config I2C_NRF5_GPIO_SCL_PIN
+config I2C_STM32
+	bool "STM32 MCU I2C Driver"
+	depends on SOC_FAMILY_STM32
+        depends on I2C_STM32_V1 || I2C_STM32_V2
```
fix whitespace

ldts replied:
yes

galak commented on drivers/i2c/Kconfig:
```diff
@@ -90,19 +90,36 @@ config I2C_NRF5_GPIO_SCL_PIN
-	bool "STM32Lx MCU I2C Interrupt Support"
-	depends on I2C_STM32LX
+config I2C_STM32_V1
+	bool "STM32 V1 Driver"
```
V1 Driver - isn't really vary useful description.

ldts replied:
stm32 uses two different i2c controllers but have no names. 
how do you propose I describe it? gen1 and gen2, v1 and v2, controller1 and controller2? 
guidelines welcome please.

ydamigos replied:
The I2C CPAL (Communication Peripheral Application Library) from ST follows the same naming. The CPAL v1 is a library which provides API for I2C peripherals of STM32F4 and CPAL v2 provides API for I2C peripherals of STM32F3 devices.

galak replied:
the symbol name is fine, in the text we can do"

bool "STM32 V1 Driver - Used on F1 & F4X SoCs"
...

galak commented on drivers/i2c/Kconfig:
```diff
@@ -90,19 +90,36 @@ config I2C_NRF5_GPIO_SCL_PIN
+	  Enable I2C support on the STM32 F1 and F4X family of processors
+
+config I2C_STM32_V2
+	bool "STM32 V2 Driver"
```
V2 Driver - isn't really vary useful description.

galak commented on drivers/i2c/Makefile:
```diff
@@ -7,5 +7,7 @@ obj-$(CONFIG_I2C_NRF5) += i2c_nrf5.o
 obj-$(CONFIG_I2C_QMSI_SS) += i2c_qmsi_ss.o
 obj-$(CONFIG_I2C_SBCON) += i2c_sbcon.o
-obj-$(CONFIG_I2C_STM32LX) += i2c_stm32lx.o
+obj-$(CONFIG_I2C_STM32_V2) += i2c_ll_stm32_v2.o i2c_ll_stm32.o
```
Have this bee:

obj-$(CONFIG_I2C_STM32) += i2c_ll_stm32.o
obj-$(CONFIG_I2C_STM32_v1) += i2c_ll_stm32_v1.o
obj-$(CONFIG_I2C_STM32_v2) += i2c_ll_stm32_v2.o

galak commented on drivers/i2c/Makefile:
```diff
@@ -7,5 +7,7 @@ obj-$(CONFIG_I2C_NRF5) += i2c_nrf5.o
 obj-$(CONFIG_I2C_SBCON) += i2c_sbcon.o
-obj-$(CONFIG_I2C_STM32LX) += i2c_stm32lx.o
+obj-$(CONFIG_I2C_STM32_V2) += i2c_ll_stm32_v2.o i2c_ll_stm32.o
+obj-$(CONFIG_I2C_STM32_V1) += i2c_ll_stm32_v1.o i2c_ll_stm32.o
```
Put V1 first

ldts replied:
missed that one. will do in the next patchset once the driver is reviewed

galak commented on drivers/i2c/Kconfig:
```diff
@@ -90,19 +90,36 @@ config I2C_NRF5_GPIO_SCL_PIN
-config I2C_STM32LX
-	bool "STM32Lx MCU I2C Driver"
-	depends on SOC_FAMILY_STM32 && SOC_SERIES_STM32L4X
+config I2C_STM32
```
Is there are 2 different drivers, remove I2C_STM32, doesn't add any particular value.

ldts replied:
true

ldts commented on ext/hal/st/stm32cube/stm32f4xx/drivers/include/stm32f4xx_hal_conf.h:
```diff
@@ -71,7 +71,6 @@
 #define HAL_SDRAM_MODULE_ENABLED
 #define HAL_HASH_MODULE_ENABLED
 #define HAL_GPIO_MODULE_ENABLED
-#define HAL_I2C_MODULE_ENABLED
```
No other STM32 platform enables the i2c LL and the HAL. When it happens, the multiple definitions of  I2C_SPEED_STANDARD and I2X_SPEED_FAST occur.

galak reviewed (changes requested):
Various things needing fixing up.

ydamigos committed:
stm32cube: build stm32f1xx_ll_i2c if CONFIG_I2C

Build stm32f1xx_ll_i2c when you enable I2C support for the stm32f1 family.

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

ydamigos committed:
drivers:i2c:i2c_ll_stm32: add support for STM32F1

Enable support for STM32F1 family.
Tested on Olimexino-STM32 board.

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

ydamigos committed:
pinmux: stm32f1x: Add I2C1 and I2C2 pins

Add I2C1 pins on PB6, PB7, PB8, PB9
Add I2C2 pins on PB10, PB11

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

ydamigos committed:
pinmux: stm32: olimexino_stm32: Add support for I2C

Add I2C support for Olimexino-STM32

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

ydamigos commented:
I validated also the v1 i2c controller on STM32F1 family.

galak commented on drivers/i2c/Kconfig:
```diff
@@ -90,19 +90,26 @@ config I2C_NRF5_GPIO_SCL_PIN
+
+config I2C_STM32_V2
+	bool "STM32 V2 Driver (F3/F7/L4X)"
 	depends on SOC_FAMILY_STM32 && SOC_SERIES_STM32L4X
```
is the depends correct if we can use this on F3 & F7 as well?

ydamigos replied:
wouldn't it be better if we change the depends when we add the i2c support (pinmux driver changes, soc pinmux changes etc) on the rest of the families?

galak replied:
That the help & 'bool' should only reference what is supported.

galak commented on drivers/i2c/Kconfig:
```diff
@@ -90,19 +90,26 @@ config I2C_NRF5_GPIO_SCL_PIN
-	bool "STM32Lx MCU I2C Driver"
+config I2C_STM32_V1
+	bool "STM32 V1 Driver (F1/L1/F4X)"
+	depends on SOC_FAMILY_STM32 && SOC_SERIES_STM32F4X
```
Is the depends correct for L1 & F1?

ydamigos replied:
wouldn't it be better if we change the depends when we add the i2c support (pinmux driver changes, soc pinmux changes etc) on the rest of the families?

galak replied:
That the help & 'bool' should only reference what is supported.

galak commented on drivers/i2c/Makefile:
```diff
@@ -7,5 +7,7 @@ obj-$(CONFIG_I2C_NRF5) += i2c_nrf5.o
+obj-$(CONFIG_I2C_STM32_V1) += i2c_ll_stm32_v1.o i2c_ll_stm32.o
+obj-$(CONFIG_I2C_STM32_V2) += i2c_ll_stm32_v2.o i2c_ll_stm32.o
 obj-$(CONFIG_TWIHS_SAM) += twihs_sam.o
+
```
remove extra whitespace

galak reviewed (changes requested):
Still need to look at the core of the driver, but some other things need tweaking as well.

galak commented on drivers/i2c/i2c_ll_stm32_v1.c:
```diff
@@ -0,0 +1,377 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * I2C Driver for: STM32F1 and STM32F4
```
What about L1?

ldts replied:
ok

erwango replied:
and F2 :-)

galak commented on drivers/i2c/i2c_ll_stm32.h:
```diff
@@ -0,0 +1,61 @@
+do { p++; len--; } while (0)
+
+s32_t msg_write(struct device *dev, struct i2c_msg *msg, u32_t flags,
+	 u16_t saddr);
```
fix line wrap to lineup with struct on previous line.

ldts replied:
will you fix checkpatch.pl to catch this?

galak replied:
not sure, I doubt it does.

galak commented on drivers/i2c/i2c_ll_stm32.h:
```diff
@@ -0,0 +1,61 @@
+s32_t msg_write(struct device *dev, struct i2c_msg *msg, u32_t flags,
+	 u16_t saddr);
+s32_t msg_read(struct device *dev, struct i2c_msg *msg, u32_t flags,
+	u16_t saddr);
```
see above line wrap comment

galak commented on drivers/i2c/i2c_ll_stm32.h:
```diff
@@ -0,0 +1,61 @@
+s32_t msg_read(struct device *dev, struct i2c_msg *msg, u32_t flags,
+	u16_t saddr);
+s32_t i2c_configure_timing(struct device *dev, u32_t clock);
+void i2c_stm32_ev_isr(void *args);
```
can we rename i2c_stm32_ev_isr to i2c_stm32_event_isr()

and i2c_stm32_er_isr -> i2c_stm32_error_isr()

ldts replied:
um...well if you insist ok

galak commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -0,0 +1,233 @@
+
+#define OPERATION(msg) (((struct i2c_msg *) msg)->flags & I2C_MSG_RW_MASK)
+
+static int i2c_stm32_transfer(struct device *dev, struct i2c_msg *message,
```
message -> msg
messages -> num_msgs

ldts replied:
:) well....ok....

jamike committed:
arm: soc: stm32: f3: add MPU capability

Since not all socs from f3 series (i.e stm32f334x8 no MPU) have MPU
capability, add capability only for MPU capable socs in Kconfig.soc

JIRA: ZEP-2220

Signed-off-by: Michel Jaouen <michel.jaouen@st.com>

galak commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -0,0 +1,233 @@
+	LL_I2C_Enable(i2c);
+
+	current = message;
+	while (messages > 0) {
```
I think I for loop is easier to understand:
for (int i = 0; i < num_msgs; i++, msg++)
{

...
}

ldts replied:
I dont agree. please let me know if is a must or a recommended

ydamigos replied:
the while loop is clear and easy to understand, I don't see why we need to change it with a for loop.

galak commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -0,0 +1,233 @@
+		if (current->len > 255)
+			return -EINVAL;
+
+		if (messages > 1) {
```
probably should have a comment here about what is going on.

ldts replied:
ok.

galak reviewed (changes requested):
Various changes to i2c_ll_stm32.c

MaureenHelm committed:
dts: Rename k64f-uart to kinetis-uart

Multiple Kinetis SoCs have the same uart hardware as the k64 and can use
the same mcux driver, so rename the dts to be more generic.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
dts: Rename kw41z-lpuart to kinetis-lpuart

Multiple Kinetis SoCs have the same lpuart hardware as the kw41z and can
use the same mcux driver, so rename the dts to be more generic.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
dts: Rename k64f-pinmux to kinetis-pinmux

Multiple Kinetis SoCs have the same pinmux hardware as the k64 and can
use the same mcux driver, so rename the dts to be more generic.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
dts: Rename k64f-gpio to kinetis-gpio

Multiple Kinetis SoCs have the same gpio hardware as the k64 and can use
the same mcux driver, so rename the dts to be more generic.

Also fixes some stranded references to kw41z-gpio to the new
kinetis-gpio.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
dts: Rename k64sim to nxp,k64f-sim

Renames k64sim to nxp,k64f-sim to be more consistent with other files.
The sim hardware can vary across Kinetis SoCs, so this dts is not made
to be generic.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

erwango committed:
pinmux: stm32f3x: add support for I2C

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
pinmux: stm32: nucleo_f334r8: Add support for I2C

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

ldts committed:
stm32: Kconfigs: select default I2C driver version

This commit also updates the board dependencies associated to this
change.

Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez-ortiz@linaro.org>

galak commented:
if you can rebase on latest branch that would be great.

erwango commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -0,0 +1,316 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * I2C Driver for: STM32F3, STM32F7 and STM32LX
```
F0, F3, F7, L0 and L4

erwango commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -191,6 +191,8 @@ static void i2c_stm32_irq_config_func_2(struct device *dev)
 
 #endif /* CONFIG_I2C_2 */
 
+#ifndef SOC_SERIES_STM32F1X
```
I2C_X availability  depends on SoC and not family.
Use #ifdef I2C3_BASE, defined in ext/hal/st/stm32cube/stm32xxxx/soc/stm32fxxxx.c files, and hence on a SoC version basis, to avoid troubles

erwango commented on drivers/i2c/Kconfig:
```diff
@@ -91,19 +91,20 @@ config I2C_NRF5_GPIO_SCL_PIN
 	default n
 	help
-	  Enable I2C support on the STM32 F3, F7 and L4X family of processors
+	  Enable I2C support on the STM32 F3 and L4X family of processors
```
dito

erwango commented on drivers/i2c/Kconfig:
```diff
@@ -91,19 +91,20 @@ config I2C_NRF5_GPIO_SCL_PIN
 	default n
 	help
-	  Enable I2C support on the STM32 F1, L1 and F4X family of processors
+	  Enable I2C support on the STM32 F1 and F4X family of processors
```
I guess this modification is a merge issue. Please keep other series information

erwango commented on drivers/i2c/Kconfig:
```diff
@@ -91,19 +91,20 @@ config I2C_NRF5_GPIO_SCL_PIN
-	bool "STM32 V2 Driver (F3/F7/L4X)"
-	depends on SOC_FAMILY_STM32 && SOC_SERIES_STM32L4X
+	bool "STM32 V2 Driver (F3/L4X)"
+	depends on SOC_FAMILY_STM32
```
dito

erwango commented on drivers/pinmux/stm32/pinmux_board_nucleo_f334r8.c:
```diff
@@ -19,6 +19,11 @@ static const struct pin_config pinconf[] = {
 	{STM32_PIN_PA10, STM32F3_PINMUX_FUNC_PA10_USART1_RX},
 #endif	/* CONFIG_UART_STM32_PORT_1 */
+#ifdef CONFIG_I2C_1
+	/* I2C2 is used for NFC, STSAFE, ToF & MEMS sensors */
```
Please remove comment (copy/paste issue)

erwango commented on arch/arm/soc/st_stm32/stm32f1/Kconfig.series:
```diff
@@ -14,5 +14,6 @@ config SOC_SERIES_STM32F1X
 	select HAS_STM32CUBE
 	select CPU_HAS_SYSTICK
 	select CLOCK_CONTROL_STM32_CUBE if CLOCK_CONTROL
+       select I2C_STM32_V1 if I2C
```
Based on this, remove CONFIG_I2C_STM32_V1 from board _defconfig

erwango commented on arch/arm/soc/st_stm32/stm32f3/Kconfig.series:
```diff
@@ -15,5 +15,6 @@ config SOC_SERIES_STM32F3X
 	select CPU_HAS_SYSTICK
 	select HAS_STM32CUBE
 	select CLOCK_CONTROL_STM32_CUBE if CLOCK_CONTROL
+       select I2C_STM32_V2 if I2C
```
dito

erwango commented on arch/arm/soc/st_stm32/stm32f4/Kconfig.series:
```diff
@@ -16,5 +16,6 @@ config SOC_SERIES_STM32F4X
 	select CPU_HAS_MPU
 	select CPU_HAS_SYSTICK
 	select CLOCK_CONTROL_STM32_CUBE if CLOCK_CONTROL
+       select I2C_STM32_V1 if I2C
```
dito

erwango commented on arch/arm/soc/st_stm32/stm32l4/Kconfig.series:
```diff
@@ -14,5 +14,6 @@ config SOC_SERIES_STM32L4X
 	select SYS_POWER_LOW_POWER_STATE_SUPPORTED
 	select CPU_HAS_SYSTICK
 	select CLOCK_CONTROL_STM32_CUBE if CLOCK_CONTROL
+       select I2C_STM32_V2 if I2C
```
dito

erwango reviewed (changes requested):
Some changes, but looks good in general

ldts commented on drivers/pinmux/stm32/pinmux_stm32f4.h:
```diff
@@ -22,6 +22,7 @@
 #define STM32F4_PINMUX_FUNC_PA3_USART2_RX	STM32_PINMUX_FUNC_ALT_7
 
 #define STM32F4_PINMUX_FUNC_PA8_UART7_RX	STM32_PINMUX_FUNC_ALT_8
+#define STM32F4_PINMUX_FUNC_PA8_I2C3_SCL	STM32_PINMUX_FUNC_ALT_4
```
ok done

dbkinder commented on boards/arm/stm32_min_dev/doc/stm32_min_dev.rst:
```diff
@@ -0,0 +1,114 @@
+Additional device support is left for the user to implement.
+
+More information on hooking up peripherals and lengthy how to articles can be
+found at `EmbedJoural`_.
```
PR https://github.com/zephyrproject-rtos/zephyr/pull/565 (on the arm branch) fixes issues with this .rst file

dbkinder reviewed (changes requested):
Merged PR #565 (on the arm branch) fixes issues with stm32_min_def.rst (rebase?)
